File: algorithms/python/15/main.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        nums.sort()
        for i in range(len(nums) - 2):
            init = nums[i]
            left, right = i + 1, len(nums) - 1
            while left < right:
                if init + nums[left] + nums[right] == 0:
                    res.add((init, nums[left], nums[right]))
                if init + nums[left] + nums[right] > 0:
                    right -= 1
                elif init + nums[left] + nums[right] <= 0:
                    left += 1
        return list(res)

File: algorithms/java/15/main.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for(int i = 0; i < nums.length - 2; i++){
            int left = i + 1, right = nums.length - 1;
            if (i > 0 && nums[i] == nums[i - 1]){
                continue;
            }
            while (left < right){
                int sum = nums[i] + nums[left] + nums[right];
                if (sum < 0){
                    left++;
                } else if (sum > 0){
                    right--;
                } else {
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left + 1]){
                        left++;
                    }
                    while (left < right && nums[right] == nums[right - 1]){
                        right--;
                    }
                    left++;
                    right--;
                }
            }
        }
        return res;
    }
}